Software/Development/SemesterProject/App/referenceHandler.c

```diff
@@ -0,0 +1,317 @@
+        return 1;
+    }
+    else    return 0;
+}
+
+int referenceTypeIsChanged(void)
+{
+    return (GUISignals.ReferenceTypePushbutton || referenceTypeHasBeenPressed());
+}
+
+/*
+ * This FSM analyses what is the reference source selected and then acquires
+ * the torque reference from Analog read or graphical user interface.
+ */
+float getReferenceValue(void)
+{
+    float analogValueTorque = 0, torqueReferenceBeforeLimit = 0;
+    switch (referenceSource)
+    {
+    case interfacePCB:
+    {
+        analogValueTorque = getTorqueReferenceSliderMeasurement();
+        torqueReferenceBeforeLimit = CalculateTorqueRefFromAnalog(analogValueTorque);
+    }break;
+    case goKart:
+    {
+        analogValueTorque = getTorqueReferencePedalMeasurement();
+        torqueReferenceBeforeLimit = CalculateTorqueRefFromAnalog(analogValueTorque);
+    }break;
+    case GUI:
+    {
+        torqueReferenceBeforeLimit = GUISignals.TorqueReference;
+    }break;
+    }
+    return torqueReferenceBeforeLimit;
+}
+
+float CalculateTorqueRefFromAnalog(float analogRead)
+{
+    return analogRead/MAX_VALUE_ADC*MAX_TORQUE_REF_NM;
+}
+
+float calculateLimitedTorqueReference(float torqueReferenceBeforeLimit)
+{
+    if (referenceIncreased(torqueReferenceBeforeLimit)) return torqueReference + deltaTorque;
+    if (referenceDecreased(torqueReferenceBeforeLimit)) return torqueReference - deltaTorque;
+    else    return torqueReference;
+}
+
+int referenceIncreased(float torqueReferenceBeforeLimit)
+{
+    return torqueReferenceBeforeLimit > torqueReference + deltaTorque;
+}
+
+int referenceDecreased(float torqueReferenceBeforeLimit)
+{
+    return torqueReferenceBeforeLimit < torqueReference - deltaTorque;
+}
+
+float calculateSaturatedTorque(float torqueBeforeSaturation)
+{
+    if (torqueBeforeSaturation > MAX_TORQUE_REF_NM) return MAX_TORQUE_REF_NM;
+    else return torqueBeforeSaturation;
+}
+
+void calculateSpeedReference(void)
+{
+    int speedReferenceBeforeLimit = 0, speedReferenceBeforeSaturation = 0;
+    speedReferenceBeforeLimit = getSpeedReferenceValue();
+    speedReferenceBeforeSaturation = calculateLimitedSpeedReference(speedReferenceBeforeLimit);
+    speedReference = calculateSaturatedSpeed(speedReferenceBeforeSaturation);
+}
+int getSpeedReferenceValue(void)
+{
+    if (speedMustBeIncreased())     undampedSpeedReference += SPEED_STEP_FROM_BUTTON_RPM;
+    if (speedMustBeDecreased())     undampedSpeedReference -= SPEED_STEP_FROM_BUTTON_RPM;
+
+    return undampedSpeedReference;
+}
+
+int calculateLimitedSpeedReference(int speedReferenceBeforeLimit)
+{
+    if (speedRefIncreased(speedReferenceBeforeLimit)) return speedReferenceBeforeLimit + deltaSpeed;
+    if (speedRefDecreased(speedReferenceBeforeLimit)) return speedReferenceBeforeLimit - deltaSpeed;
+    else return speedReferenceBeforeLimit;
+}
+
+int speedRefIncreased(int speedReferenceBeforeLimit)
```

Comment:
Maybe i'm missing something but it seems that with this logic, in some moments it might never reach the desired speed.
Lets say the delta is 10, speedReferenceBeforeLimit is 45 and speedReference is 40. Then this function returns a 0 and speed is never increased all the way to 45 but instead it stays at 40.
Maybe the issue is solved somewhere else but i havent been able to see it yet.

Reply:
A way to solve this issue could be to write another if statement such that if (speedReferenceBeforeLimit < speedReference) speedReference = speedReferenceBeforeLimit;
That is to be done in the previous function i suppose.

Reply:
You are right that it will never reach the exact value, but deltas are user customisable, so for example getting the default values of 5 RPM, the error would be 0.024 km/h if my calculations are correct. Do you think it's a big deal? I think it is not and even if the number were correct, I don't think our whole system will allow such high accuracy.

Reply:
Ok that makes sense, wouldn't it have some influence in the control since it doesn't reach the expected value? IDK, just give it a thought but i dont think it's a big deal either.

Reply:
I think is a fair concern. I will think around. Thanks!